PushProjectionIntoTableScan  fails with: "Multiple entries with same key"

When I upgraded PreSTO338 to 344,Here's what happened;
Caused by: java.lang.IllegalArgumentException: Multiple entries with same key: @2452f4c3: null=null and @2452f4c3: null=null
	at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:215)
	at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:209)
	at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:146)
	at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:109)
	at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:394)
	at io.prestosql.sql.planner.iterative.rule.PushProjectionIntoTableScan.apply(PushProjectionIntoTableScan.java:149)
	at io.prestosql.sql.planner.iterative.rule.PushProjectionIntoTableScan.apply(PushProjectionIntoTableScan.java:56)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.transform(IterativeOptimizer.java:168)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreNode(IterativeOptimizer.java:143)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:115)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.optimize(IterativeOptimizer.java:99)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:206)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:195)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:190)
	at io.prestosql.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:444)
	at io.prestosql.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:424)
	at io.prestosql.execution.SqlQueryExecution.start(SqlQueryExecution.java:376)
	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:245)
	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:139)
	at io.prestosql.$gen.Presto_346____20201204_011754_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

--- a/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushProjectionIntoTableScan.java
+++ b/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushProjectionIntoTableScan.java
@@ -98,28 +98,20 @@ public class PushProjectionIntoTableScan
     {
         TableScanNode tableScan = captures.get(TABLE_SCAN);
 
-        Map<Symbol, Expression> inputExpressions = project.getAssignments().getMap();
-
-        ImmutableList.Builder<NodeRef<Expression>> nodeReferencesBuilder = ImmutableList.builder();
-        ImmutableList.Builder<ConnectorExpression> partialProjectionsBuilder = ImmutableList.builder();
-
         // Extract translatable components from projection expressions. Prepare a mapping from these internal
         // expression nodes to corresponding ConnectorExpression translations.
-        for (Map.Entry<Symbol, Expression> expression : inputExpressions.entrySet()) {
-            Map<NodeRef<Expression>, ConnectorExpression> partialTranslations = extractPartialTranslations(
-                    expression.getValue(),
-                    context.getSession(),
-                    typeAnalyzer,
-                    context.getSymbolAllocator().getTypes());
+        Map<NodeRef<Expression>, ConnectorExpression> partialTranslations = project.getAssignments().getMap().entrySet().stream()
+                .flatMap(expression ->
+                        extractPartialTranslations(
+                                expression.getValue(),
+                                context.getSession(),
+                                typeAnalyzer,
+                                context.getSymbolAllocator().getTypes()).entrySet().stream())
+                // Avoid duplicates
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue, (first, ignore) -> first));
 
-            partialTranslations.forEach((nodeRef, expr) -> {
-                nodeReferencesBuilder.add(nodeRef);
-                partialProjectionsBuilder.add(expr);
-            });
-        }
-
-        List<NodeRef<Expression>> nodesForPartialProjections = nodeReferencesBuilder.build();
-        List<ConnectorExpression> connectorPartialProjections = partialProjectionsBuilder.build();
+        List<NodeRef<Expression>> nodesForPartialProjections = ImmutableList.copyOf(partialTranslations.keySet());
+        List<ConnectorExpression> connectorPartialProjections = ImmutableList.copyOf(partialTranslations.values());
 
         Map<String, Symbol> inputVariableMappings = tableScan.getAssignments().keySet().stream()
                 .collect(toImmutableMap(Symbol::getName, identity()));

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestConnectorPushdownRulesWithHive.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestConnectorPushdownRulesWithHive.java
@@ -46,7 +46,6 @@ import io.trino.spi.type.RowType;
 import io.trino.spi.type.Type;
 import io.trino.spi.type.TypeOperators;
 import io.trino.sql.planner.Symbol;
-import io.trino.sql.planner.assertions.PlanMatchPattern;
 import io.trino.sql.planner.iterative.rule.PruneTableScanColumns;
 import io.trino.sql.planner.iterative.rule.PushPredicateIntoTableScan;
 import io.trino.sql.planner.iterative.rule.PushProjectionIntoTableScan;
@@ -280,7 +279,7 @@ public class TestConnectorPushdownRulesWithHive
                 })
                 .matches(
                         strictProject(
-                                ImmutableMap.of("expr", PlanMatchPattern.expression("COLA")),
+                                ImmutableMap.of("expr", expression("COLA")),
                                 tableScan(
                                         hiveTable.withProjectedColumns(ImmutableSet.of(columnA))::equals,
                                         TupleDomain.all(),

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestConnectorPushdownRulesWithHive.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestConnectorPushdownRulesWithHive.java
@@ -53,6 +53,9 @@ import io.trino.sql.planner.iterative.rule.PushProjectionIntoTableScan;
 import io.trino.sql.planner.iterative.rule.test.BaseRuleTest;
 import io.trino.sql.planner.iterative.rule.test.PlanBuilder;
 import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.tree.ArithmeticBinaryExpression;
+import io.trino.sql.tree.ArithmeticUnaryExpression;
+import io.trino.sql.tree.Expression;
 import io.trino.sql.tree.LongLiteral;
 import io.trino.sql.tree.SubscriptExpression;
 import io.trino.sql.tree.SymbolReference;
@@ -78,6 +81,8 @@ import static io.trino.sql.planner.assertions.PlanMatchPattern.filter;
 import static io.trino.sql.planner.assertions.PlanMatchPattern.project;
 import static io.trino.sql.planner.assertions.PlanMatchPattern.strictProject;
 import static io.trino.sql.planner.assertions.PlanMatchPattern.tableScan;
+import static io.trino.sql.tree.ArithmeticBinaryExpression.Operator.ADD;
+import static io.trino.sql.tree.ArithmeticUnaryExpression.Sign.MINUS;
 import static io.trino.testing.TestingConnectorSession.SESSION;
 import static io.trino.testing.TestingSession.testSessionBuilder;
 import static java.lang.String.format;
@@ -280,6 +285,87 @@ public class TestConnectorPushdownRulesWithHive
                                         hiveTable.withProjectedColumns(ImmutableSet.of(columnA))::equals,
                                         TupleDomain.all(),
                                         ImmutableMap.of("COLA", columnA::equals))));
+
+        metastore.dropTable(new HiveIdentity(SESSION), SCHEMA_NAME, tableName, true);
+    }
+
+    @Test
+    public void testPushdownWithDuplicateExpressions()
+    {
+        String tableName = "duplicate_expressions";
+        tester().getQueryRunner().execute(format(
+                "CREATE TABLE  %s (struct_of_bigint, just_bigint) AS SELECT cast(row(5, 6) AS row(a bigint, b bigint)) AS struct_of_int, 5 AS just_bigint WHERE false",
+                tableName));
+
+        PushProjectionIntoTableScan pushProjectionIntoTableScan = new PushProjectionIntoTableScan(
+                tester().getMetadata(),
+                tester().getTypeAnalyzer(),
+                new ScalarStatsCalculator(tester().getMetadata(), tester().getTypeAnalyzer()));
+
+        HiveTableHandle hiveTable = new HiveTableHandle(SCHEMA_NAME, tableName, ImmutableMap.of(), ImmutableList.of(), ImmutableList.of(), Optional.empty());
+        TableHandle table = new TableHandle(new CatalogName(HIVE_CATALOG_NAME), hiveTable, new HiveTransactionHandle(), Optional.empty());
+
+        HiveColumnHandle bigintColumn = createBaseColumn("just_bigint", 1, toHiveType(BIGINT), BIGINT, REGULAR, Optional.empty());
+        HiveColumnHandle partialColumn = new HiveColumnHandle(
+                "struct_of_bigint",
+                0,
+                toHiveType(ROW_TYPE),
+                ROW_TYPE,
+                Optional.of(new HiveColumnProjectionInfo(
+                        ImmutableList.of(0),
+                        ImmutableList.of("a"),
+                        toHiveType(BIGINT),
+                        BIGINT)),
+                REGULAR,
+                Optional.empty());
+
+        // Test projection pushdown with duplicate column references
+        tester().assertThat(pushProjectionIntoTableScan)
+                .on(p -> {
+                    SymbolReference column = p.symbol("just_bigint", BIGINT).toSymbolReference();
+                    Expression negation = new ArithmeticUnaryExpression(MINUS, column);
+                    return p.project(
+                            Assignments.of(
+                                    // The column reference is part of both the assignments
+                                    p.symbol("column_ref", BIGINT), column,
+                                    p.symbol("negated_column_ref", BIGINT), negation),
+                            p.tableScan(
+                                    table,
+                                    ImmutableList.of(p.symbol("just_bigint", BIGINT)),
+                                    ImmutableMap.of(p.symbol("just_bigint", BIGINT), bigintColumn)));
+                })
+                .matches(project(
+                        ImmutableMap.of(
+                                "column_ref", expression("just_bigint_0"),
+                                "negated_column_ref", expression("- just_bigint_0")),
+                        tableScan(
+                                hiveTable.withProjectedColumns(ImmutableSet.of(bigintColumn))::equals,
+                                TupleDomain.all(),
+                                ImmutableMap.of("just_bigint_0", bigintColumn::equals))));
+
+        // Test Dereference pushdown
+        tester().assertThat(pushProjectionIntoTableScan)
+                .on(p -> {
+                    SubscriptExpression subscript = new SubscriptExpression(p.symbol("struct_of_bigint", ROW_TYPE).toSymbolReference(), new LongLiteral("1"));
+                    Expression sum = new ArithmeticBinaryExpression(ADD, subscript, new LongLiteral("2"));
+                    return p.project(
+                            Assignments.of(
+                                    // The subscript expression instance is part of both the assignments
+                                    p.symbol("expr_deref", BIGINT), subscript,
+                                    p.symbol("expr_deref_2", BIGINT), sum),
+                            p.tableScan(
+                                    table,
+                                    ImmutableList.of(p.symbol("struct_of_bigint", ROW_TYPE)),
+                                    ImmutableMap.of(p.symbol("struct_of_bigint", ROW_TYPE), partialColumn.getBaseColumn())));
+                })
+                .matches(project(
+                        ImmutableMap.of(
+                                "expr_deref", expression(new SymbolReference("struct_of_bigint#a")),
+                                "expr_deref_2", expression(new ArithmeticBinaryExpression(ADD, new SymbolReference("struct_of_bigint#a"), new LongLiteral("2")))),
+                        tableScan(
+                                hiveTable.withProjectedColumns(ImmutableSet.of(partialColumn))::equals,
+                                TupleDomain.all(),
+                                ImmutableMap.of("struct_of_bigint#a", partialColumn::equals))));
 
         metastore.dropTable(new HiveIdentity(SESSION), SCHEMA_NAME, tableName, true);
     }

